moveit_ros/trajectory_cache/src/utils/utils.cpp

```diff
@@ -0,0 +1,659 @@
+                                           const MoveGroupInterface& move_group, double match_tolerance,
+                                           const std::string& reference_frame_id, const std::string& prefix)
+{
+  const std::shared_ptr<tf2_ros::Buffer> tf_buffer = move_group.getTF();  // NOLINT: Deliberate lifetime extension.
+
+  // Make ignored members explicit.
+
+  bool emit_position_constraint_warning = false;
+  for (auto& constraint : constraints)
+  {
+    for (auto& position_constraint : constraint.position_constraints)
+    {
+      if (!position_constraint.constraint_region.primitives.empty())
+      {
+        emit_position_constraint_warning = true;
+        break;
+      }
+    }
+    if (emit_position_constraint_warning)
+    {
+      break;
+    }
+  }
+  if (emit_position_constraint_warning)
+  {
+    RCLCPP_WARN_STREAM(getLogger(), "Ignoring " << prefix << ".position_constraints.constraint_region: Not supported.");
+  }
+
+  bool emit_visibility_constraint_warning = false;
+  for (auto& constraint : constraints)
+  {
+    if (!constraint.visibility_constraints.empty())
+    {
+      emit_visibility_constraint_warning = true;
+      break;
+    }
+  }
+  if (emit_visibility_constraint_warning)
+  {
+    RCLCPP_WARN_STREAM(getLogger(), "Ignoring " << prefix << ".visibility_constraints: Not supported.");
+  }
+
+  // Begin extraction.
+
+  size_t constraint_idx = 0;
+  for (auto& constraint : constraints)
+  {
+    // We sort to avoid cardinality.
+    sortJointConstraints(constraint.joint_constraints);
+    sortPositionConstraints(constraint.position_constraints);
+    sortOrientationConstraints(constraint.orientation_constraints);
+
+    std::string constraint_prefix = prefix + "_" + std::to_string(constraint_idx++);
+
+    // Joint constraints
+    size_t joint_idx = 0;
+    for (auto& joint_constraint : constraint.joint_constraints)
+    {
+      std::string meta_name = constraint_prefix + ".joint_constraints_" + std::to_string(joint_idx++);
+      query.append(meta_name + ".joint_name", joint_constraint.joint_name);
+      queryAppendCenterWithTolerance(query, meta_name + ".position", joint_constraint.position, match_tolerance);
+      query.appendGTE(meta_name + ".tolerance_above", joint_constraint.tolerance_above);
+      query.appendLTE(meta_name + ".tolerance_below", joint_constraint.tolerance_below);
+    }
+
+    // Position constraints
+    if (!constraint.position_constraints.empty())
+    {
+      // All offsets will be "frozen" and computed wrt. the workspace frame instead.
+      query.append(constraint_prefix + ".position_constraints.header.frame_id", reference_frame_id);
+
+      size_t position_idx = 0;
+      for (auto& position_constraint : constraint.position_constraints)
+      {
+        std::string meta_name = constraint_prefix + ".position_constraints_" + std::to_string(position_idx++);
+
+        // Compute offsets wrt. to workspace frame.
+        double x_offset = 0;
```

Comment:
Ah, this transform block repeats here. So +1 to my request to factor this into a helper.